864 hi l1b compute coincidence type and time deltas

tmplummer opened the pull request:
# Change Summary

## Overview
Adds the Hi L1B algorithm for computing coincidence type and delta_t values. These fields share enough commonality in computation that I decided to implement them in a single function.

## Updated Files
- imap_processing/hi/l1b/hi_l1b.py
   - Added function for computing L1B coincidence_type and delta_t variables
   - Coincidence type and delta_t variables are now added to the dataset within that function.
- imap_processing/hi/utils.py
   - Added hi constants frozen dataclass similar to how other instruments have implemented
- imap_processing/tests/hi/test_hi_l1b.py
   - Added fixture that generates synthetic L1A data for testing coincidence_type and delta_t algorithm
   - Added test coverage for compute_coincidence_type_and_time_deltas function

Closes: #864

tmplummer committed:
Add hi_l1b function for computing coincidence type and time deltas for direct events

tmplummer committed:
Add test for compute coincidence type and time deltas funcion

tmplummer committed:
Add test coverage for coincidence type and delta T computation

tmplummer committed:
Cleanup for PR

tmplummer committed:
Fix broken doc build

sdhoyt commented on imap_processing/hi/l1b/hi_l1b.py:
```diff
@@ -90,28 +107,149 @@ def annotate_direct_events(l1a_dataset: xr.Dataset) -> xr.Dataset:
+    # |      2      |      B      |      A      |      C1     |      C2     |
+    # |      3      |      C1     |      A      |      B      |      C2     |
+    # Set coincidence type bitmask
+    out_ds.coincidence_type[a_mask | tof_1_valid_mask] |= CoincidenceBitmap.A.value
```
I think because you're using an `IntEnum` you can do `CoincidenceBitmap.A` instead of `CoincidenceBitmap.A.value`

sdhoyt commented on imap_processing/hi/l1b/hi_l1b.py:
```diff
@@ -90,28 +107,149 @@ def annotate_direct_events(l1a_dataset: xr.Dataset) -> xr.Dataset:
+
+    # Table denoting how hit-first mask and valid TOF masks are used to set
+    # coincidence type bitmask
+    # -----------------------------------------------------------------------
```
I like this table, but I'm a little confused how this relates to the code below. For line 189 it looks like the coincidence type is being set  to the A value for all indices where either `a_mask` or `tof_1_valid_mask` are true. But I'm not sure how to get that from the table.

tmplummer replied:
I wondered if that implementation would cause confusion. The union of events where `a_mask == True` with events where `a_mask == False & tof_1_valid_mask == True` is the same as the union of events where `a_mask == True` with events where `tof_1_valid_mask == True` because the events in the `a_mask == True` set contain both `a_mask == True & tof_1_valid_mask == True` and `a_mask == True & tof_1_valid_mask == False`. I thought I would be clever and make the code more succinct, but perhaps it just makes it more confusing?

tmplummer replied:
Hopefully, adding both coincidence type bits implied by valid TOF values to the table made it easier to see? It made me catch a simplification to the C1 mask.

tech3371 replied:
I think the setup is great. It would be helpful if you can step through this in today's meeting. Quick overview.

sdhoyt commented on imap_processing/tests/hi/test_hi_l1b.py:
```diff
@@ -29,3 +37,117 @@ def test_hi_l1b_de(create_de_data, tmp_path):
+def synthetic_trigger_id_and_tof_data():
+    """Create synthetic minimum dataset for testing the
+    coincidence_type_and_time_deltas algorithm."""
+    # The following coincidence type table shows possible values to consider
```
This table is really helpful! I should add something similar for Lo's coincidence types

sdhoyt commented on imap_processing/tests/hi/test_hi_l1b.py:
```diff
@@ -29,3 +37,117 @@ def test_hi_l1b_de(create_de_data, tmp_path):
+    tof2s = np.array(np.concatenate((HiConstants.TOF2_BAD_VALUES, [2])))
+    tof3s = np.array(np.concatenate((HiConstants.TOF3_BAD_VALUES, [3])))
+    var_names = ["trigger_id", "tof_1", "tof_2", "tof_3"]
+    data = np.meshgrid(ids, tof1s, tof2s, tof3s)
```
Great idea to more easily cover the combinations!

tmplummer replied:
I can't quite take credit... found it on SO.

sdhoyt reviewed (commented):
Looks great! Just hoping to get some clarification on how to read those tables in the comments

tmplummer committed:
Clearify coincidence type table
Simplify C1 coincidence type mask logic
Remove .value from IntEnums where not needed

laspsandoval commented on docs/source/conf.py:
```diff
@@ -107,6 +107,8 @@
     (r"py:.*", r".*CoDICECompression.*"),
     (r"py:.*", r".*.lo.l0.utils.*"),
     (r"py:.*", r".*.lo.l0.data_classes.*"),
```
What is this for?

tmplummer replied:
The documentation build fails without this. The sphinx inheritance stuff seems like it is a mess and so sometimes, you just have to add things here to ignore the warnings.

laspsandoval commented on imap_processing/tests/hi/test_hi_l1b.py:
```diff
@@ -29,3 +37,109 @@ def test_hi_l1b_de(create_de_data, tmp_path):
+def test_compute_coincidence_type_and_time_deltas(synthetic_trigger_id_and_tof_data):
+    """Test coverage for
+    `imap_processing.hi.hi_l1b.compute_coincidence_type_and_time_deltas`."""
+    # l1a_dataset = load_cdf(hi_l1a_test_file_path)
```
I think you meant to delete this?

laspsandoval commented on imap_processing/hi/l1b/hi_l1b.py:
```diff
@@ -90,28 +110,149 @@ def annotate_direct_events(l1a_dataset: xr.Dataset) -> xr.Dataset:
+    out_ds.coincidence_type[
+        b_mask | (a_mask & tof_1_valid_mask) | (c_mask & tof_2_valid_mask)
+    ] |= CoincidenceBitmap.B
+    out_ds.coincidence_type[c_mask | tof_2_valid_mask | tof_2_valid_mask] |= (
```
I think that you can simplify:
out_ds.coincidence_type[c_mask | tof_2_valid_mask] |= CoincidenceBitmap.C1.value

laspsandoval commented on imap_processing/hi/l1b/hi_l1b.py:
```diff
@@ -90,28 +110,149 @@ def annotate_direct_events(l1a_dataset: xr.Dataset) -> xr.Dataset:
+        -out_ds.tof_1.values[b_mask & tof_1_valid_mask].astype(np.float32)
+        * HiConstants.TOF1_TICK_PER_NS
+    )
+    out_ds.delta_t_ab.values[c_mask & tof_1and2_valid_mask] = (
```
I think that this could be split into two separate steps for clarity:

delta_t_2 = out_ds.tof_2.values[c_mask & tof_1and2_valid_mask].astype(np.float32) * HiConstants.TOF2_TICK_PER_NS
delta_t_1 = out_ds.tof_1.values[c_mask & tof_1and2_valid_mask].astype(np.float32) * HiConstants.TOF1_TICK_PER_NS
out_ds.delta_t_ab.values[c_mask & tof_1and2_valid_mask] = delta_t_2 - delta_t_1

laspsandoval commented on imap_processing/hi/l1b/hi_l1b.py:
```diff
@@ -90,28 +110,149 @@ def annotate_direct_events(l1a_dataset: xr.Dataset) -> xr.Dataset:
+        out_ds.tof_2.values[a_mask & tof_2_valid_mask].astype(np.float32)
+        * HiConstants.TOF2_TICK_PER_NS
+    )
+    out_ds.delta_t_ac1.values[b_mask & tof_1and2_valid_mask] = (
```
I think this could be split for clarity:

delta_t_2 = out_ds.tof_2.values[b_mask & tof_1and2_valid_mask] * HiConstants.TOF2_TICK_PER_NS
delta_t_1 = out_ds.tof_1.values[b_mask & tof_1and2_valid_mask] * HiConstants.TOF1_TICK_PER_NS
out_ds.delta_t_ac1.values[b_mask & tof_1and2_valid_mask] = delta_t_2 - delta_t_1

laspsandoval commented on imap_processing/hi/l1b/hi_l1b.py:
```diff
@@ -90,28 +110,149 @@ def annotate_direct_events(l1a_dataset: xr.Dataset) -> xr.Dataset:
+    )
+
+    # delta_t_bc1
+    out_ds.delta_t_bc1.values[a_mask & tof_1_valid_mask & tof_2_valid_mask] = (
```
Same here I would split it up:
delta_t_2 = out_ds.tof_2.values[a_mask & tof_1and2_valid_mask] * HiConstants.TOF2_TICK_PER_NS
delta_t_1 = out_ds.tof_1.values[a_mask & tof_1and2_valid_mask] * HiConstants.TOF1_TICK_PER_NS
out_ds.delta_t_bc1.values[a_mask & tof_1_valid_mask & tof_2_valid_mask] = delta_t_2 - delta_t_1

laspsandoval commented on imap_processing/hi/l1b/hi_l1b.py:
```diff
@@ -90,28 +110,149 @@ def annotate_direct_events(l1a_dataset: xr.Dataset) -> xr.Dataset:
+    # |      2      |      B      |     A,B     |     B,C1    |    C1,C2    |
+    # |      3      |      C1     |     A,C1    |     B,C1    |    C1,C2    |
+    # Set coincidence type bitmask
+    out_ds.coincidence_type[a_mask | tof_1_valid_mask] |= CoincidenceBitmap.A
```
For me personally I would really appreciate a docstring on each bitmask. Maybe it's overkill for some people, but it helps me to immediately understand the purpose.

Example:
Set the 'A' bit in the 'coincidence_type' variable where:
- The trigger is 'A' (using `a_mask`), OR
- `tof_1` is valid (using `tof_1_valid_mask`).

tmplummer replied:
My philosophy is that if you need a bunch of comments to explain your code, then the code is not written in a way that it can be clearly understood. IMO, the comments you wrote are exactly what the code says with a slightly different order. For this instance, I think the comment is redundant.

laspsandoval reviewed (approved):
I think that programmatically everything looks in order. My suggestion would be to add more docstrings, but you can take it or leave it.

tmplummer committed:
Address PR comments
Rework some of the code to make it more readable

tmplummer commented:
> I think that programmatically everything looks in order. My suggestion would be to add more ~~docstrings~~ comments, but you can take it or leave it.

@laspsandoval, I took your feedback about needing more comments to mean that my code was not easily understandable. I made some minor tweaks that make it much more readable, IMO. I also added some comments and think that it should be very clear now.

sdhoyt reviewed (approved):
Thanks for the walkthrough of all this. I think everything looks good!
